### DISCUSSION TITLE
External Boundary Conditions

### DISCUSSION THREAD

**Discussion by alahyari** [AMReX-Codes/amrex#2266] [Q&A] [answered]
Hi all,

I'm trying to implement external boundary conditions, and I'm facing a rather strange problem. Let's say I have external BCs on all my boundaries, so I set `amrex::BCRec bcs` to `amrex::BCType::ext_dir` for all components in all directions. Then, according to the instructions, I wrote the following function to fill the boundary cell:
```
void external_bc(amrex::Box const& bx, 
	       amrex::Array4<amrex::Real> const& q,
	       const int dcomp, 
	       const int numcomp,
	       amrex::GeometryData const& geom, 
	       const amrex::Real time,
	       const amrex::BCRec* bcr, 
	       const int bcomp, 
	       const int orig_comp)
{
    \* check bcr values in lower end and higher end of each direction and 
       if equal to "amrex::BCType::ext_dir", fill the boundary cells accordingly */
}
```
at which the pointer function `bcfill` of type `UserFillBox` points to, which is then passed to an `amrex::PhysBCFunct` object of type `amrex::CpuBndryFuncFab` for filling the patches (see below for lev=0).
```
if (lev == 0)
    {
        // current level multifab and time in case for sub-time-stepping
        amrex::Vector<amrex::MultiFab*> smf;
        amrex::Vector<amrex::Real> stime;
        get_data(0, time, smf, stime);

        amrex::CpuBndryFuncFab bndry_func(bcfill);
        amrex::PhysBCFunct<amrex::CpuBndryFuncFab> physbc(geom[lev], bcs, bndry_func);

        amrex::FillPatchSingleLevel(mf, time, smf, stime, 0, 0, nv, geom[lev], physbc, 0);
    }
```
So far, everything's good. But, when `external_bc` function is called, the `const amrex::BCRec* bcr` does not seem to be consistent with the original `amrex::BCRec bcs`. When I check the `bcr` values, it seems only the lower end of each direction is `amrex::BCType::ext_dir` and the higher ends are `amrex::BCType::int_dir` for some reason. I double checked the `bcs` values right before being passed to the `amrex::PhysBCFunct` object and all is fine. Am I missing anything? Has anyone encountered a similar issue? 

Thank you in advance.

**Comment by WeiqunZhang** [comment, marked as answer]
I think everything is fine.  The `bcr` passed to your `external_bc` is not the original `BCRec` for the whole domain.  It's the BC for that box.  For boxes with higher ends not touching the physical domain, the value there will not be external dirichlet.

**Reply by alahyari** [reply]
Oh, gosh! That does make sense. Thank you @WeiqunZhang
